apps/omg_child_chain/lib/omg_child_chain/block_queue/submission_monitor.ex

```diff
@@ -0,0 +1,174 @@
+      Enum.filter(state.pending_blocks, fn {_blknum, first_submit_height} ->
+        state.root_chain_height - first_submit_height >= state.stall_threshold_blocks
+      end)
+
+    _ = :telemetry.execute([:blocks_stalled, __MODULE__], %{blocks: stalled_blocks})
+    _ = log_stalled_blocks(stalled_blocks, state.root_chain_height)
+    _ = trigger_alarm(state.alarm_module, state.alarm_raised, stalled_blocks)
+
+    {:noreply, state}
+  end
+
+  # Keeps track of the latest root chain height
+  def handle_info({:internal_event_bus, :ethereum_new_height, new_height}, state) do
+    {:noreply, %{state | root_chain_height: new_height}}
+  end
+
+  # Listens for a block being submitted and add it to monitoring if it hasn't been tracked
+  def handle_info({:internal_event_bus, :block_submitting, blknum}, state) do
+    pending_blocks = add_new_blknum(state.pending_blocks, blknum, state.root_chain_height)
+    {:noreply, %{state | pending_blocks: pending_blocks}}
+  end
+
+  # Listens for a block that got submitted and drop it from monitoring
+  def handle_info({:internal_event_bus, :block_submitted, blknum}, state) do
+    pending_blocks = remove_blknum(state.pending_blocks, blknum)
+    {:noreply, %{state | pending_blocks: pending_blocks}}
+  end
+
+  # Ignore unrelated events
+  def handle_info({:internal_event_bus, :enqueue_block, _}, state) do
+    {:noreply, state}
+  end
+
+  #
+  # Handle incoming alarms
+  #
+  # These functions are called by the AlarmHandler so that this monitor process can update
+  # its internal state according to the raised alarms.
+  #
+  def handle_cast({:set_alarm, :block_submit_stalled}, state) do
+    {:noreply, %{state | alarm_raised: true}}
+  end
+
+  def handle_cast({:clear_alarm, :block_submit_stalled}, state) do
+    {:noreply, %{state | alarm_raised: false}}
+  end
+
+  #
+  # Private functions
+  #
+
+  # Add the blknum to tracking only if it is not already tracked
+  @spec add_new_blknum([{blknum(), any()}], blknum(), non_neg_integer()) :: [pending_block()]
+  defp add_new_blknum(pending_blocks, blknum, root_chain_height) do
+    case Enum.any?(pending_blocks, fn {pending_blknum, _} -> pending_blknum == blknum end) do
+      true -> pending_blocks
+      false -> [{blknum, root_chain_height} | pending_blocks]
+    end
+  end
+
+  @spec remove_blknum([{blknum(), any()}], blknum()) :: [pending_block()]
+  defp remove_blknum(pending_blocks, blknum) do
+    Enum.reject(pending_blocks, fn {pending_blknum, _} -> pending_blknum == blknum end)
+  end
+
+  #
+  # Alarms management
+  #
+
+  defp install_alarm_handler() do
+    case Enum.member?(:gen_event.which_handlers(:alarm_handler), __MODULE__.AlarmHandler) do
+      true -> :ok
+      _ -> :alarm_handler.add_alarm_handler(__MODULE__.AlarmHandler)
+    end
+  end
+
+  @spec trigger_alarm(module(), boolean(), [blknum()]) :: :ok
+  defp trigger_alarm(_alarm_module, false, []), do: :ok
+
+  defp trigger_alarm(alarm_module, false, _stalled_blocks) do
```

Comment:
why `state.alarm_raised` false should set alarm while true we clear? Should it be the opposite?

Reply:
The 2nd parameter is whether the alarm was already raised, and when `_stalled_blocks` is non-zero, it will be caught by this case.

But I see it's a valid confusion, updated the clauses to this instead https://github.com/omgnetwork/elixir-omg/pull/1649/commits/d676020cda4efa555e2e376f811cc063e5a78020:

```elixir
  @spec trigger_alarm(alarm :: module(), alarm_raised :: boolean(), stalled_blocks :: [blknum()]) :: :ok
  defp trigger_alarm(alarm_module, false, stalled_blocks) when length(stalled_blocks) >= 1 do
    alarm_module.set(alarm_module.block_submit_stalled(__MODULE__))
  end

  defp trigger_alarm(alarm_module, true, []) do
    alarm_module.clear(alarm_module.block_submit_stalled(__MODULE__))
  end

  defp trigger_alarm(_alarm_module, _alarm_raised, _stalled_blocks), do: :ok
```

Reply:
I see